[3019] Restore support for default and studio-related actions and templates

sbegaudeau opened the pull request:
- [fix] Add support for images as static resources
- [3019] Restore support for default and studio-related actions and templates

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [ ] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

## Review

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

sbegaudeau committed:
[3019] Restore support for default and studio-related actions and templates

Bug: https://github.com/eclipse-sirius/sirius-web/issues/3019
Signed-off-by: Stéphane Bégaudeau <stephane.begaudeau@obeo.fr>

sbegaudeau committed:
[fix] Add support for images as static resources

Signed-off-by: Stéphane Bégaudeau <stephane.begaudeau@obeo.fr>

pcdavid commented:
I'll have a look when this is no longer marked as a draft.

sbegaudeau commented:
> I'll have a look when this is no longer marked as a draft.

It's still not ready for review but I'm getting closer, in the first version I was missing the ImageController to return the images of the project templates and the domain and view metamodels were never provided in an editing context creating for a studio. Now one of the last thing I need to handle is the execution of editing context actions from the onboard area (I'm at least missing a data fetcher). Pretty soon, you'll be able to use the frontend and trigger the real editing context actions from the onboard area (new domain / new view model).

It will not provide full support for domain and view DSLs but it should allow us to create studios and edit domain and view models "soon".

pcdavid commented on packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/studio/services/StudioProjectTemplateInitializer.java:
```diff
@@ -0,0 +1,93 @@
+/*******************************************************************************
+ * Copyright (c) 2024, 2024 Obeo.
```
Nitpick: there are several instances of "2024, 2024" in different files.

sbegaudeau replied:
I've fixed a dozens or so of those, thanks 👍

sbegaudeau committed:
[3019] Restore support for portal representations

Bug: https://github.com/eclipse-sirius/sirius-web/issues/3019
Signed-off-by: Stéphane Bégaudeau <stephane.begaudeau@obeo.fr>

sbegaudeau committed:
[test] Add missing projects to the global test coverage

Signed-off-by: Stéphane Bégaudeau <stephane.begaudeau@obeo.fr>

sbegaudeau commented:
Now that I'm confident (since I have a test to prove it) that I can also open representations such as portals, this PR can be reviewed. I've also added some projects that were missing from the global test coverage and updated the check-coverage script to ensure that pretty much everybody is equally included in it (I have not included the soon to be removed existing sirius web projects for example).

I'm now quite confident that, as a next step, I'll be able to add the remaining representations fairly quickly.

pcdavid commented on CHANGELOG.adoc:
```diff
@@ -27,6 +27,7 @@
 - The type `RenameProjectSuccessPayload` and `DeletedProjectSuccessPayload` will be replaced by `SuccessPayload`
 - The type `DeleteRepresentationSuccessPayload` will be replaced by `SuccessPayload`
 - The mutation `createRepresentation` and `deleteRepresentation` and their datafetcher will be entirely managed by Sirius Components
+- The interface `IEditingContextEPackageService` should be deleted since we now have access to editing context preprocessors.
```
The link is not clear, to me at least. `IEditingContextEPackageService` has a single `getEPackages` method; what is the relation with pre-processors?

sbegaudeau replied:
I've added some additional information

pcdavid commented on packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/api/ITemplateBasedProjectCreator.java:
```diff
@@ -0,0 +1,27 @@
+ * @author sbegaudeau
+ */
+public interface ITemplateBasedProjectCreator {
+    IPayload createProjectFromTemplate(CreateProjectFromTemplateInput input, UUID projectId, IProjectTemplateInitializer initializer);
```
Isn't it strange to passe the initializer as an argument here? The input contains the `templateId` which determines the initializer to use.

sbegaudeau replied:
Yes, I've moved the retrieval of the initializer closer to where it is used.

pcdavid commented on packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/TemplateBasedProjectCreator.java:
```diff
@@ -0,0 +1,77 @@
+
+    @Override
+    @Transactional
+    public IPayload createProjectFromTemplate(CreateProjectFromTemplateInput input, UUID projectId, IProjectTemplateInitializer initializer) {
```
Not sure about the naming here. `ITemplateBasedProjectCreator.createProjectFromTemplate`, but the project has actually already been created when this is called. This is more a project content initializer or something like that.
It's also very close to an editing context action in what it can do; it's just invoked at a different time (right after the project is created) and from a different UI mechanism (from buttons in the projects page and not exposed once the project is created/opened).

sbegaudeau replied:
I was not fan of the name either, I'll change it.

pcdavid commented on packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/configurationproperties/PortalDisabledTests.java:
```diff
@@ -0,0 +1,57 @@
+    private IEditingContextSearchService editingContextSearchService;
+
+    @Test
+    @DisplayName("Given all Sirius Web features enabled, when representation descriptions are requested, then portal support is available")
```
The display name does not match what is tested, i.e. portal not available when the feature is disabled.

sbegaudeau replied:
Indeed, I'll fix that

pcdavid commented on packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/EditingContextControllerIntegrationTests.java:
```diff
@@ -126,12 +154,37 @@ public void givenEditingContextIdWhenQueryIsPerformedThenItsActionsAreReturned()
+
+        StepVerifier.create(flux)
+                .then(invokeEditingContextActionTask)
+                .thenCancel()
```
Isn't this missing some "expectNextMatches"?

sbegaudeau replied:
No it's on purpose here since nothing will be sent on this subscription when we subscribed to it. Even if something was sent, it's not relevant for my need. I just need the `EditingContextEventProcessor` to be loaded to perform the mutation, I don't really care about the events returned by the `EditingContextEventProcessor` here.

pcdavid commented on packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/semanticdata/listeners/SemanticDataInitializer.java:
```diff
@@ -38,6 +38,6 @@ public SemanticDataInitializer(ISemanticDataCreationService semanticDataCreation
     @Transactional
     @TransactionalEventListener
```
`@Transactional(propagation = Propagation.REQUIRES_NEW)` required so that other listeners can react to the `SemanticDataCreatedEvent` send during this.

sbegaudeau commented:
It should be better now :)
